### ISSUE TITLE
Network type clarification

### ISSUE THREAD

**Issue by SilverWingedSeraph** [unbase/unbase#5]
If I understand correctly, the Network struct is meant to be a wrapper around the simulator so the simulator can be swapped out for a real backend, like TCP/IP or raw datagrams or packet radio or pidgeons with USB drives or whatever. If that's the case, it might be better to use a trait object there; given a trait `NetworkBackend`:

```
#[derive(Clone)]
pub struct Network {
    shared: Arc<NetworkShared>,
    backend: Box<NetworkBackend>
}
```

This would make the implementation backend-agnostic. I proposed `Box` because I'd imagine that the Network would own its backend, but it's possible that an `Arc` would be more appropriate, if the backend could be shared.

**Comment by dnorman** [comment]
Your understanding is mostly correct. The Network struct is intended to be the facilitator of channels between slabs. There might (in the future) be a significant number of channel modules. For Slabs which are co-resident in the same OS process, these channels should be actual mpsc channels w/o a middle-man, and wrapped as thinly as possible for efficiency purposes. In the cases of remote Slabs, they'll be something pluggable for UDP, WebRTP, etc. In many cases, these channels may have to be implemented external to the library, for instance: Once compilation to javascript is in play, presumably the WebRTP portion will have to be implemented in javascript and made available via dependency injection. 

The resident slab-to-slab communications module probably ought to be available by default, with all other channel modules being injected. The test suite needs only override the default slab-to-slab direct channels to ensure determinism, and avoid injecting any other channel modules.

Presumably the default co-resident slab-to-slab channel module could itself be injected too, merely substituted with the simulator in the case of the test suite.

I'm still sharpening my rust knowledge/skills, so the suggestions and PRs are more than welcome. Keep 'em coming :)

**Comment by SilverWingedSeraph** [comment]
I think a trait object is definitely the way to go if you expect to use DI - it just gives the most flexibility. The only question is whether to use a `Box` or an `Arc`, and that really comes down to whether the same channel will need to be shared between multiple `Network`s.

**Comment by dnorman** [comment]
I don't have any plans to share channel modules ( I was thinking the term "channel" should be reserved for each slab-to-slab connection, or at least for each recipient slab, but I could be convinced otherwise )

My plan is to Trait-ify the channel modules, and inject them into the network struct so as to add capabilities. The network struct would then call upon each channel module to generate a channel corresponding to each destination slab as appropriate/needed.

It occurred to me somewhat belatedly that the main reason I had used an Arc is that I need to be able to interact with the simulator inside the test suite, thus requiring that both the test suite and the network have access to it. I'm not averse to switching the channel module to use a box, and then Arc-ify the Simulator module internally, allowing for clones. (Been taking a shoot-first-ask-questions-later approach to try to get the bones in place. This will of course entail much tidying later.)

Speaking of the simulator:
The most recent head scratcher is how to set up the simulator so it's robust, yet easy to use.
Given that cross-slab scenarios presently require blocking operations, this requires the requesting slab be in it's own thread when calling subject.get_value(). This is because there would otherwise be a deadlock in cases which requires requesting a memo from a remote slab, and waiting for a response from said slab.

My thought is that I'd employ futures for the simulator (more confident about the acceptability of that rather than futures in the internals). Presumably it'd be something like:

simulator.run_as_new_thread_and_advance_clock_in_present_thread_as_needed_until_return(|| {
    //closure here doing the blocking call that requires a request and response from another slab
}).run_as_new_thread_and_advance_clock_in_present_thread_as_needed_until_return(|| {
   //closure for the next thing
}).then_something_else()

I haven't fully wrapped my head around the implications here, and I've only tinkered with futures in rust thus far, so suggestions/design-patterns welcomed.
